// @aws-cdk/aws-glue-alpha/lib/data-quality-ruleset.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var _a, _b;
var DataQualityRuleset_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataQualityRuleset = exports.DataQualityTargetTable = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdk = require("aws-cdk-lib");
const core_1 = require("aws-cdk-lib/core");
const aws_glue_1 = require("aws-cdk-lib/aws-glue");
const metadata_resource_1 = require("aws-cdk-lib/core/lib/metadata-resource");
const prop_injectable_1 = require("aws-cdk-lib/core/lib/prop-injectable");
/**
 * Properties of a DataQualityTargetTable.
 */
class DataQualityTargetTable {
    constructor(databaseName, tableName) {
        this.databaseName = databaseName;
        this.tableName = tableName;
    }
}
exports.DataQualityTargetTable = DataQualityTargetTable;
_a = JSII_RTTI_SYMBOL_1;
DataQualityTargetTable[_a] = { fqn: "@aws-cdk/aws-glue-alpha.DataQualityTargetTable", version: "2.202.0-alpha.0" };
/**
 * A Glue Data Quality ruleset.
 */
let DataQualityRuleset = DataQualityRuleset_1 = class DataQualityRuleset extends core_1.Resource {
    static fromRulesetArn(scope, id, rulesetArn) {
        class Import extends core_1.Resource {
            constructor() {
                super(...arguments);
                this.rulesetArn = rulesetArn;
                this.rulesetName = cdk.Arn.extractResourceName(rulesetArn, 'dataqualityruleset');
            }
        }
        return new Import(scope, id);
    }
    static fromRulesetName(scope, id, rulesetName) {
        class Import extends core_1.Resource {
            constructor() {
                super(...arguments);
                this.rulesetArn = DataQualityRuleset_1.buildRulesetArn(scope, rulesetName);
                this.rulesetName = rulesetName;
            }
        }
        return new Import(scope, id);
    }
    static buildRulesetArn(scope, rulesetName) {
        return cdk.Stack.of(scope).formatArn({
            service: 'glue',
            resource: 'dataqualityruleset',
            resourceName: rulesetName,
        });
    }
    constructor(scope, id, props) {
        super(scope, id, {
            physicalName: props.rulesetName,
        });
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_DataQualityRulesetProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, DataQualityRuleset);
            }
            throw error;
        }
        // Enhanced CDK Analytics Telemetry
        (0, metadata_resource_1.addConstructMetadata)(this, props);
        const rulesetResource = new aws_glue_1.CfnDataQualityRuleset(this, 'Resource', {
            clientToken: props.clientToken,
            description: props.description,
            name: props.rulesetName,
            ruleset: props.rulesetDqdl,
            tags: props.tags,
            targetTable: props.targetTable,
        });
        const resourceName = this.getResourceNameAttribute(rulesetResource.ref);
        this.rulesetArn = DataQualityRuleset_1.buildRulesetArn(this, resourceName);
        this.rulesetName = resourceName;
    }
};
exports.DataQualityRuleset = DataQualityRuleset;
_b = JSII_RTTI_SYMBOL_1;
DataQualityRuleset[_b] = { fqn: "@aws-cdk/aws-glue-alpha.DataQualityRuleset", version: "2.202.0-alpha.0" };
/** Uniquely identifies this class. */
DataQualityRuleset.PROPERTY_INJECTION_ID = '@aws-cdk.aws-glue-alpha.DataQualityRuleset';
exports.DataQualityRuleset = DataQualityRuleset = DataQualityRuleset_1 = __decorate([
    prop_injectable_1.propertyInjectable
], DataQualityRuleset);
