// @aws-cdk/aws-glue-alpha/lib/storage-parameter.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageParameter = exports.StorageParameters = exports.OrcColumnMappingType = exports.WriteParallel = exports.ColumnCountMismatchHandlingAction = exports.SurplusCharHandlingAction = exports.SurplusBytesHandlingAction = exports.NumericOverflowHandlingAction = exports.InvalidCharHandlingAction = exports.CompressionType = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * The compression type.
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_ > _"compression_type"_
 */
var CompressionType;
(function (CompressionType) {
    /**
     * No compression.
     */
    CompressionType["NONE"] = "none";
    /**
     * Burrows-Wheeler compression.
     */
    CompressionType["BZIP2"] = "bzip2";
    /**
     * Deflate compression.
     */
    CompressionType["GZIP"] = "gzip";
    /**
     * Compression algorithm focused on high compression and decompression speeds, rather than the maximum possible compression.
     */
    CompressionType["SNAPPY"] = "snappy";
})(CompressionType || (exports.CompressionType = CompressionType = {}));
/**
 * Specifies the action to perform when query results contain invalid UTF-8 character values.
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_ > _"invalid_char_handling"_
 */
var InvalidCharHandlingAction;
(function (InvalidCharHandlingAction) {
    /**
     * Doesn't perform invalid character handling.
     */
    InvalidCharHandlingAction["DISABLED"] = "DISABLED";
    /**
     * Cancels queries that return data containing invalid UTF-8 values.
     */
    InvalidCharHandlingAction["FAIL"] = "FAIL";
    /**
     * Replaces invalid UTF-8 values with null.
     */
    InvalidCharHandlingAction["SET_TO_NULL"] = "SET_TO_NULL";
    /**
     * Replaces each value in the row with null.
     */
    InvalidCharHandlingAction["DROP_ROW"] = "DROP_ROW";
    /**
     * Replaces the invalid character with the replacement character you specify using `REPLACEMENT_CHAR`.
     */
    InvalidCharHandlingAction["REPLACE"] = "REPLACE";
})(InvalidCharHandlingAction || (exports.InvalidCharHandlingAction = InvalidCharHandlingAction = {}));
/**
 * Specifies the action to perform when ORC data contains an integer (for example, BIGINT or int64) that is larger than the column definition (for example, SMALLINT or int16).
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_ > _"numeric_overflow_handling"_
 */
var NumericOverflowHandlingAction;
(function (NumericOverflowHandlingAction) {
    /**
     * Invalid character handling is turned off.
     */
    NumericOverflowHandlingAction["DISABLED"] = "DISABLED";
    /**
     * Cancel the query when the data includes invalid characters.
     */
    NumericOverflowHandlingAction["FAIL"] = "FAIL";
    /**
     * Set invalid characters to null.
     */
    NumericOverflowHandlingAction["SET_TO_NULL"] = "SET_TO_NULL";
    /**
     * Set each value in the row to null.
     */
    NumericOverflowHandlingAction["DROP_ROW"] = "DROP_ROW";
})(NumericOverflowHandlingAction || (exports.NumericOverflowHandlingAction = NumericOverflowHandlingAction = {}));
/**
 * Specifies how to handle data being loaded that exceeds the length of the data type defined for columns containing VARBYTE data. By default, Redshift Spectrum sets the value to null for data that exceeds the width of the column.
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_ > _"surplus_bytes_handling"_
 */
var SurplusBytesHandlingAction;
(function (SurplusBytesHandlingAction) {
    /**
     * Replaces data that exceeds the column width with null.
     */
    SurplusBytesHandlingAction["SET_TO_NULL"] = "SET_TO_NULL";
    /**
     * Doesn't perform surplus byte handling.
     */
    SurplusBytesHandlingAction["DISABLED"] = "DISABLED";
    /**
     * Cancels queries that return data exceeding the column width.
     */
    SurplusBytesHandlingAction["FAIL"] = "FAIL";
    /**
     * Drop all rows that contain data exceeding column width.
     */
    SurplusBytesHandlingAction["DROP_ROW"] = "DROP_ROW";
    /**
     * Removes the characters that exceed the maximum number of characters defined for the column.
     */
    SurplusBytesHandlingAction["TRUNCATE"] = "TRUNCATE";
})(SurplusBytesHandlingAction || (exports.SurplusBytesHandlingAction = SurplusBytesHandlingAction = {}));
/**
 * Specifies how to handle data being loaded that exceeds the length of the data type defined for columns containing VARCHAR, CHAR, or string data. By default, Redshift Spectrum sets the value to null for data that exceeds the width of the column.
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_ > _"surplus_char_handling"_
 */
var SurplusCharHandlingAction;
(function (SurplusCharHandlingAction) {
    /**
     * Replaces data that exceeds the column width with null.
     */
    SurplusCharHandlingAction["SET_TO_NULL"] = "SET_TO_NULL";
    /**
     * Doesn't perform surplus character handling.
     */
    SurplusCharHandlingAction["DISABLED"] = "DISABLED";
    /**
     * Cancels queries that return data exceeding the column width.
     */
    SurplusCharHandlingAction["FAIL"] = "FAIL";
    /**
     * Replaces each value in the row with null.
     */
    SurplusCharHandlingAction["DROP_ROW"] = "DROP_ROW";
    /**
     * Removes the characters that exceed the maximum number of characters defined for the column.
     */
    SurplusCharHandlingAction["TRUNCATE"] = "TRUNCATE";
})(SurplusCharHandlingAction || (exports.SurplusCharHandlingAction = SurplusCharHandlingAction = {}));
/**
 * Identifies if the file contains less or more values for a row than the number of columns specified in the external table definition. This property is only available for an uncompressed text file format.
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_ > _"column_count_mismatch_handling"_
 */
var ColumnCountMismatchHandlingAction;
(function (ColumnCountMismatchHandlingAction) {
    /**
     * Column count mismatch handling is turned off.
     */
    ColumnCountMismatchHandlingAction["DISABLED"] = "DISABLED";
    /**
     * Fail the query if the column count mismatch is detected.
     */
    ColumnCountMismatchHandlingAction["FAIL"] = "FAIL";
    /**
     * Fill missing values with NULL and ignore the additional values in each row.
     */
    ColumnCountMismatchHandlingAction["SET_TO_NULL"] = "SET_TO_NULL";
    /**
     * Drop all rows that contain column count mismatch error from the scan.
     */
    ColumnCountMismatchHandlingAction["DROP_ROW"] = "DROP_ROW";
})(ColumnCountMismatchHandlingAction || (exports.ColumnCountMismatchHandlingAction = ColumnCountMismatchHandlingAction = {}));
/**
 * Specifies how to handle data being loaded that exceeds the length of the data type defined for columns containing VARCHAR, CHAR, or string data. By default, Redshift Spectrum sets the value to null for data that exceeds the width of the column.
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_ > _"surplus_char_handling"_
 */
var WriteParallel;
(function (WriteParallel) {
    /**
     * Write data in parallel.
     */
    WriteParallel["ON"] = "on";
    /**
     * Write data serially.
     */
    WriteParallel["OFF"] = "off";
})(WriteParallel || (exports.WriteParallel = WriteParallel = {}));
/**
 * Specifies how to map columns when the table uses ORC data format.
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_ > _"orc.schema.resolution"_
 */
var OrcColumnMappingType;
(function (OrcColumnMappingType) {
    /**
     * Map columns by name.
     */
    OrcColumnMappingType["NAME"] = "name";
    /**
     * Map columns by position.
     */
    OrcColumnMappingType["POSITION"] = "position";
})(OrcColumnMappingType || (exports.OrcColumnMappingType = OrcColumnMappingType = {}));
/**
 * The storage parameter keys that are currently known, this list is not exhaustive and other keys may be used.
 */
var StorageParameters;
(function (StorageParameters) {
    /**
     * The number of rows to skip at the top of a CSV file when the table is being created.
     */
    StorageParameters["SKIP_HEADER_LINE_COUNT"] = "skip.header.line.count";
    /**
     * Determines whether data handling is on for the table.
     */
    StorageParameters["DATA_CLEANSING_ENABLED"] = "data_cleansing_enabled";
    /**
     * The type of compression used on the table, when the file name does not contain an extension. This value overrides the compression type specified through the extension.
     */
    StorageParameters["COMPRESSION_TYPE"] = "compression_type";
    /**
     * Specifies the action to perform when query results contain invalid UTF-8 character values.
     */
    StorageParameters["INVALID_CHAR_HANDLING"] = "invalid_char_handling";
    /**
     * Specifies the replacement character to use when you set `INVALID_CHAR_HANDLING` to `REPLACE`.
     */
    StorageParameters["REPLACEMENT_CHAR"] = "replacement_char";
    /**
     * Specifies the action to perform when ORC data contains an integer (for example, BIGINT or int64) that is larger than the column definition (for example, SMALLINT or int16).
     */
    StorageParameters["NUMERIC_OVERFLOW_HANDLING"] = "numeric_overflow_handling";
    /**
     * Specifies how to handle data being loaded that exceeds the length of the data type defined for columns containing VARBYTE data. By default, Redshift Spectrum sets the value to null for data that exceeds the width of the column.
     */
    StorageParameters["SURPLUS_BYTES_HANDLING"] = "surplus_bytes_handling";
    /**
     * Specifies how to handle data being loaded that exceeds the length of the data type defined for columns containing VARCHAR, CHAR, or string data. By default, Redshift Spectrum sets the value to null for data that exceeds the width of the column.
     */
    StorageParameters["SURPLUS_CHAR_HANDLING"] = "surplus_char_handling";
    /**
     * Identifies if the file contains less or more values for a row than the number of columns specified in the external table definition. This property is only available for an uncompressed text file format.
     */
    StorageParameters["COLUMN_COUNT_MISMATCH_HANDLING"] = "column_count_mismatch_handling";
    /**
     * A property that sets the numRows value for the table definition. To explicitly update an external table's statistics, set the numRows property to indicate the size of the table. Amazon Redshift doesn't analyze external tables to generate the table statistics that the query optimizer uses to generate a query plan. If table statistics aren't set for an external table, Amazon Redshift generates a query execution plan based on an assumption that external tables are the larger tables and local tables are the smaller tables.
     */
    StorageParameters["NUM_ROWS"] = "num_rows";
    /**
     * A property that sets number of rows to skip at the beginning of each source file.
     */
    StorageParameters["SERIALIZATION_NULL_FORMAT"] = "serialization.null.format";
    /**
     * A property that sets the column mapping type for tables that use ORC data format. This property is ignored for other data formats.
     */
    StorageParameters["ORC_SCHEMA_RESOLUTION"] = "orc.schema.resolution";
    /**
     * A property that sets whether CREATE EXTERNAL TABLE AS should write data in parallel. When 'write.parallel' is set to off, CREATE EXTERNAL TABLE AS writes to one or more data files serially onto Amazon S3. This table property also applies to any subsequent INSERT statement into the same external table.
     */
    StorageParameters["WRITE_PARALLEL"] = "write.parallel";
    /**
     * A property that sets the maximum size (in MB) of each file written to Amazon S3 by CREATE EXTERNAL TABLE AS. The size must be a valid integer between 5 and 6200. The default maximum file size is 6,200 MB. This table property also applies to any subsequent INSERT statement into the same external table.
     */
    StorageParameters["WRITE_MAX_FILESIZE_MB"] = "write.maxfilesize.mb";
    /**
     * You can specify an AWS Key Management Service key to enable Server–Side Encryption (SSE) for Amazon S3 objects.
     */
    StorageParameters["WRITE_KMS_KEY_ID"] = "write.kms.key.id";
})(StorageParameters || (exports.StorageParameters = StorageParameters = {}));
/**
 * A storage parameter. The list of storage parameters available is not exhaustive and other keys may be used.
 *
 * If you would like to specify a storage parameter that is not available as a static member of this class, use the `StorageParameter.custom` method.
 *
 * The list of storage parameters currently known within the CDK is listed.
 *
 * @see https://docs.aws.amazon.com/glue/latest/dg/table-properties-crawler.html
 *
 * @see https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_EXTERNAL_TABLE.html#r_CREATE_EXTERNAL_TABLE-parameters - under _"TABLE PROPERTIES"_
 */
class StorageParameter {
    /**
     * The number of rows to skip at the top of a CSV file when the table is being created.
     */
    static skipHeaderLineCount(value) {
        return new StorageParameter('skip.header.line.count', value.toString());
    }
    /**
     * Determines whether data handling is on for the table.
     */
    static dataCleansingEnabled(value) {
        return new StorageParameter('data_cleansing_enabled', value.toString());
    }
    /**
     * The type of compression used on the table, when the file name does not contain an extension. This value overrides the compression type specified through the extension.
     */
    static compressionType(value) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_CompressionType(value);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.compressionType);
            }
            throw error;
        }
        return new StorageParameter('compression_type', value);
    }
    /**
     * Specifies the action to perform when query results contain invalid UTF-8 character values.
     */
    static invalidCharHandling(value) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_InvalidCharHandlingAction(value);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.invalidCharHandling);
            }
            throw error;
        }
        return new StorageParameter('invalid_char_handling', value);
    }
    /**
     * Specifies the replacement character to use when you set `INVALID_CHAR_HANDLING` to `REPLACE`.
     */
    static replacementChar(value) {
        return new StorageParameter('replacement_char', value);
    }
    /**
     * Specifies the action to perform when ORC data contains an integer (for example, BIGINT or int64) that is larger than the column definition (for example, SMALLINT or int16).
     */
    static numericOverflowHandling(value) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_NumericOverflowHandlingAction(value);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.numericOverflowHandling);
            }
            throw error;
        }
        return new StorageParameter('numeric_overflow_handling', value);
    }
    /**
     * Specifies how to handle data being loaded that exceeds the length of the data type defined for columns containing VARBYTE data. By default, Redshift Spectrum sets the value to null for data that exceeds the width of the column.
     */
    static surplusBytesHandling(value) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_SurplusBytesHandlingAction(value);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.surplusBytesHandling);
            }
            throw error;
        }
        return new StorageParameter('surplus_bytes_handling', value);
    }
    /**
     * Specifies how to handle data being loaded that exceeds the length of the data type defined for columns containing VARCHAR, CHAR, or string data. By default, Redshift Spectrum sets the value to null for data that exceeds the width of the column.
     */
    static surplusCharHandling(value) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_SurplusCharHandlingAction(value);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.surplusCharHandling);
            }
            throw error;
        }
        return new StorageParameter('surplus_char_handling', value);
    }
    /**
     * Identifies if the file contains less or more values for a row than the number of columns specified in the external table definition. This property is only available for an uncompressed text file format.
     */
    static columnCountMismatchHandling(value) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_ColumnCountMismatchHandlingAction(value);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.columnCountMismatchHandling);
            }
            throw error;
        }
        return new StorageParameter('column_count_mismatch_handling', value);
    }
    /**
     * A property that sets the numRows value for the table definition. To explicitly update an external table's statistics, set the numRows property to indicate the size of the table. Amazon Redshift doesn't analyze external tables to generate the table statistics that the query optimizer uses to generate a query plan. If table statistics aren't set for an external table, Amazon Redshift generates a query execution plan based on an assumption that external tables are the larger tables and local tables are the smaller tables.
     */
    static numRows(value) {
        return new StorageParameter('num_rows', value.toString());
    }
    /**
     * A property that sets number of rows to skip at the beginning of each source file.
     */
    static serializationNullFormat(value) {
        return new StorageParameter('serialization.null.format', value);
    }
    /**
     * A property that sets the column mapping type for tables that use ORC data format. This property is ignored for other data formats. If this property is omitted, columns are mapped by `OrcColumnMappingType.NAME` by default.
     *
     * @default OrcColumnMappingType.NAME
     */
    static orcSchemaResolution(value) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_OrcColumnMappingType(value);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.orcSchemaResolution);
            }
            throw error;
        }
        return new StorageParameter('orc.schema.resolution', value);
    }
    /**
     * A property that sets whether CREATE EXTERNAL TABLE AS should write data in parallel. When 'write.parallel' is set to off, CREATE EXTERNAL TABLE AS writes to one or more data files serially onto Amazon S3. This table property also applies to any subsequent INSERT statement into the same external table.
     *
     * @default WriteParallel.ON
     */
    static writeParallel(value) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_glue_alpha_WriteParallel(value);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.writeParallel);
            }
            throw error;
        }
        return new StorageParameter('write.parallel', value);
    }
    /**
     * A property that sets the maximum size (in MB) of each file written to Amazon S3 by CREATE EXTERNAL TABLE AS. The size must be a valid integer between 5 and 6200. The default maximum file size is 6,200 MB. This table property also applies to any subsequent INSERT statement into the same external table.
     */
    static writeMaxFileSizeMb(value) {
        return new StorageParameter('write.maxfilesize.mb', value.toString());
    }
    /**
     * You can specify an AWS Key Management Service key to enable Server–Side Encryption (SSE) for Amazon S3 objects.
     */
    static writeKmsKeyId(value) {
        return new StorageParameter('write.kms.key.id', value);
    }
    /**
     * A custom storage parameter.
     * @param key - The key of the storage parameter.
     * @param value - The value of the storage parameter.
     */
    static custom(key, value) {
        return new StorageParameter(key, value.toString());
    }
    constructor(key, value) {
        this.key = key;
        this.value = value;
    }
}
exports.StorageParameter = StorageParameter;
_a = JSII_RTTI_SYMBOL_1;
StorageParameter[_a] = { fqn: "@aws-cdk/aws-glue-alpha.StorageParameter", version: "2.202.0-alpha.0" };
